include/libpmemobj++/array_iterator.hpp

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright 2018, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Iterators for pmem::obj::array
+ */
+
+#ifndef PMEMOBJ_ARRAY_ITERATOR_HPP
+#define PMEMOBJ_ARRAY_ITERATOR_HPP
+
+#include <algorithm>
+#include <functional>
+
+namespace pmem
+{
+
+namespace obj
+{
+
+/*
+ * Base class for array iterators which satisfies RandomAccessIterator
+ */
+template <typename Iterator, typename Reference, typename Pointer>
+struct base_iterator {
+	/**
+	* Constructor taking a pointer
+	*/
+	constexpr base_iterator(Pointer begin) : ptr(begin)
+	{
+	}
+
+	/**
+	* Dereference operator
+	*/
+	constexpr Reference operator*() const
+	{
+		return *ptr;
+	}
+
+	/**
+	* Arrow operator
+	*/
+	constexpr Pointer operator->() const
+	{
+		return ptr;
+	}
+
+	/**
+	* Prefix increment operator
+	*/
+	Iterator &operator++()
+	{
+		static_cast<Iterator *>(this)->change_by(1);
```

Comment:
why do we need this cast? wouldn't just `return *this;` work?
And does it have to be a `static_cast`?